bob-backend/src/pearl/disk_controller.rs

```diff
@@ -216,28 +303,28 @@ impl DiskController {
         Ok(())
     }
 
-    pub(crate) async fn run(&self) -> AnyResult<()> {
-        if *self.state.read().await == GroupsState::Initialized {
-            let res = {
-                let _permit = self.run_sem.acquire().await.expect("Semaphore is closed");
-                Self::run_groups(self.groups.clone()).await
-            };
-            if let Err(e) = res {
-                error!("Can't run groups on disk {:?} (reason: {})", self.disk, e);
-                // if work dir became unavailable we should reinit, so new state is NotReady
-                // otherwise - we'll try again
-                if !Self::is_work_dir_available(self.disk.path()) {
-                    *self.state.write().await = GroupsState::NotReady;
-                    *self.groups.write().await = Vec::new();
+    async fn groups_run(&self) -> AnyResult<()> {
+        let state = self.state.read().await.clone();
+        match state {
+            GroupsState::Initialized => {
+                let res = {
+                    let _permit = self.run_sem.acquire().await.expect("Semaphore is closed");
+                    Self::run_groups(self.groups.clone()).await
+                };
```

Comment:
map and map_err. And probably separate function to keep "match" clean

Reply:
I agree with you in part with separation, but I don't understand, how I can use `map` and `map_err` (because they get closuse, which is not `async`, but I need to perform async operations)